File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var numberOfSets = function(n, k) {
    const MOD = 1000000007;

    let dp = new Array(n).fill(1);

    for (let j = 1; j <= k; j++) {
        const next = new Array(n).fill(0);
        let sum = 0;

        for (let i = 1; i < n; i++) {
            sum = (sum + dp[i - 1]) % MOD;
            next[i] = (next[i - 1] + sum) % MOD;
        }
        dp = next;
    }
    
    return dp[n - 1];
};

File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.java
class Solution {
    public int numberOfSets(int n, int k) {
        final int MOD = 1_000_000_007;

        long[] dp = new long[n];

        // 0 segments -> 1 way for every prefix
        for (int i = 0; i < n; i++) dp[i] = 1;

        for (int j = 1; j <= k; j++) {
            long[] next = new long[n];
            long sum = 0;

            for (int i = 1; i < n; i++) {
                sum = (sum + dp[i - 1] % MOD);
                next[i] = (next[i - 1] + sum) % MOD;
            }
            dp = next;
        }
        return (int) dp[n - 1];
    }
}